distributed_shampoo/fsdp_shampoo.py

```diff
@@ -0,0 +1,849 @@
+                )
+            if beta1 != 0.0 and EXP_AVG not in state:
+                state[EXP_AVG] = torch.zeros_like(
+                    p.grad, memory_format=torch.preserve_format
+                )
+
+            # Generate split lists.
+            split_params.extend(state[PRECONDITIONERS].combine_and_split_dims(p))
+            split_momentum_directions.extend(
+                state[PRECONDITIONERS].combine_and_split_dims(state[MOMENTUM])
+                if momentum_param != 0.0
+                else []
+            )
+
+        return split_params, split_preconditioned_grads, split_momentum_directions
+
+    @torch.no_grad()
+    def _iterate_step(self) -> torch.Tensor:
+        iteration = None
+        for group in self.param_groups:
+            for p in group[PARAMS]:
+                # skip parameters not on worker
+                if p.numel() == 0:
+                    continue
+
+                self.state[p][STEP] += 1
+                iteration = self.state[p][STEP]
+        return iteration
+
+    @torch.no_grad()
+    def reset_preconditioners(self):
+        for group in self.param_groups:
+            for p in group[PARAMS]:
+                # skip parameters not on worker
+                if p.numel() == 0:
+                    continue
+
+                self.state[p][PRECONDITIONERS].reset_preconditioners()
+
+    @torch.no_grad()
+    def step(self, closure=None):
+        """Performs a single optimization step.
+
+        Args:
+            closure (callable, optional): A closure that reevaluates the model
+                and returns the loss.
+
+        """
+        loss = None
+        if closure is not None:
+            with torch.enable_grad():
+                loss = closure()
+
+        iteration = self._iterate_step()
+        self._update_preconditioners()
+
+        # for debugging
+        current_rank = dist.get_rank()
+
+        # Computes root inverse of all preconditioners every self._precondition_frequency
+        # after the self._start_preconditioning_step iteration.
+        if (
+            iteration % self._precondition_frequency == 0
+            and iteration >= self._start_preconditioning_step
+        ):
+            self._compute_root_inverse()
+
+            if self._debug_mode:
+                self._compute_and_log_root_inverse_residuals()
+
+        # Loops over all parameter groups and parameters to perform update.
+        for group in self.param_groups:
+            beta1, _ = group[BETAS]
+            momentum_param = group[MOMENTUM]
+            weight_decay = group[WEIGHT_DECAY]
+            lr = group[LR]
+
+            (
+                split_params,
+                split_preconditioned_grads,
+                split_momentum_directions,
+            ) = self._init_group(group)
+
+            for p in group[PARAMS]:
+                state = self.state[p]
+                if p.grad is None or not state[PRECONDITIONERS].on_source_rank:
```

Comment:
We can remove the `not state[PRECONDITIONERS].on_source_rank` piece since we are no longer relying on any distributed computation. Do we also want to perform a consistent check to see if `p.numel() == 0` here?
```suggestion
                if p.numel() == 0 or p.grad is None:
```